### PR TITLE
WIP FB8-144: Send events to async slaves after ACKed by at least one semi-sync slave

### PR DESCRIPTION
**Pull request by inikep**
```
Because of conflicts I removed changes in `mysql-test/t/all_persisted_variables.test`. 
This test has to be modified at `let $total_persistent_vars=XXX;` (+1 increase)
and it should be re-recorded.
```

Jira issue: https://jira.percona.com/browse/FB8-144

Reference Patch: https://github.com/facebook/mysql-5.6/commit/cc4e803
Reference Patch: https://github.com/facebook/mysql-5.6/commit/aad3ce2a54e

---------- https://github.com/facebook/mysql-5.6/commit/cc4e803 ----------

Send events to async slaves after ACKed by at least one semi-sync slave

Summary:
Added a new mysql variable `wait_for_semi_sync_ack` to control
sending binlog events to async slaves. When this variables is set,
events are sent to async slave only after it is ACKed by at least one
semi-sync slave (if any).

Originally Reviewed By: Tema

fbshipit-source-id: 9cad7e5

---------- https://github.com/facebook/mysql-5.6/commit/aad3ce2a54e ----------
rpl_semi_sync_master_crash_if_active_trxs
Fixed rpl_wait_for_semi_sync_ack

Summary:rpl_semi_sync_master_crash_if_active_trxs
Following changes are made to fix the feature
 - Condition variable is registered in dump thread's THD before waiting so that
   the thread can respond to kill command
 - rpl_wait_for_semi_sync_ack is respected only when semi-sync master is enabled
   (i.e rpl_semi_sync_master_enabled = 1)
 - Added a status variable Rpl_semi_sync_master_ack_waits which counts the
   number of times we waited for an ACK (useful for benchmarking)rpl_semi_sync_master_crash_if_active_trxs

Squash with D6022785

Originally Reviewed By: hermanlee

fbshipit-source-id: 64710f9

### CODE REVIEW

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.cnf [hunk 1]
```diff
@@ -0,0 +1,15 @@
+!include ../my.cnf
+
+[mysqld.1]
+log-slave-updates
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.cnf]: log-slave-updates is already a default in 8.0

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+set @@global.debug= '+d,before_semi_sync_reply';
+
+source include/rpl_connection_master.inc;
+send create table t1 (a int);
+
+connection semi_sync_slave;
+set debug_sync='now WAIT_FOR reached';
+
+connection async_slave;
+echo "Async Slave: Should not contain any tables";
```
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]: Here and below, consider assert.inc with I_S.TABLES queries

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+set @@global.rpl_wait_for_semi_sync_ack= true;
+set @save.rpl_semi_sync_master_timeout= @@global.rpl_semi_sync_master_timeout;
+set @@global.rpl_semi_sync_master_timeout= 10000000;
+call mtr.add_suppression("Run function 'wait_for_semi_sync_ack' in plugin 'rpl_semi_sync_master' failed");
+call mtr.add_suppression("Error while waiting for semi-sync ACK on dump thread");
+call mtr.add_suppression("A message intended for a client cannot be sent there as no client-session is attached");
+call mtr.add_suppression("Timeout waiting for reply of binlog");
+
+
+connection semi_sync_slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]: Here and everywhere
```suggestion
let $rpl_connection_name= semi_sync_slave;
source include/rpl_connect.inc;
```

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+set @@global.debug= '+d,before_semi_sync_reply';
+
+source include/rpl_connection_master.inc;
+send create table t1 (a int);
+
+connection semi_sync_slave;
+set debug_sync='now WAIT_FOR reached';
+
+connection async_slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]: Here and everywhere
```suggestion
let $rpl_connection_name= async_slave;
source include/rpl_connect.inc;
```

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+connection semi_sync_slave;
+set debug_sync='now SIGNAL continue';
+
+echo "Switching rpl_semi_sync_master_enabled back on";
+source include/rpl_connection_master.inc;
+reap;
+set @@global.rpl_semi_sync_master_enabled = 1;
+
+connection semi_sync_slave;
+stop slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]:
```suggestion
source include/stop_slave.inc
```

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+start slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]:
```suggestion
source include/start_slave.inc
```

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+connection semi_sync_slave;
+set debug_sync='now SIGNAL continue';
+
+echo "Switching rpl_wait_for_semi_sync_ack back on";
+source include/rpl_connection_master.inc;
+reap;
+set @@global.rpl_wait_for_semi_sync_ack = 1;
+
+connection semi_sync_slave;
+stop slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]:
```suggestion
source include/stop_slave.inc
```

File: mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+start slave;
```
Comment (percona-ysorokin) [inline mysql-test/suite/rpl_gtid/t/rpl_wait_for_semi_sync_ack.test]:
```suggestion
source include/start_slave.inc
```

File: plugin/semisync/semisync_master.h [hunk 9]
```diff
@@ -579,6 +582,10 @@ class ReplSemiSyncMaster : public ReplSemiSyncBase {
    */
   mysql_mutex_t LOCK_binlog_;
 
+  mysql_mutex_t LOCK_last_acked;
+  mysql_cond_t COND_last_acked;
+  std::pair<std::string, my_off_t> last_acked;
```
Comment (percona-ysorokin) [inline plugin/semisync/semisync_master.h]: typedef/using for a complex type
```suggestion
using binlog_position_type = std::pair<std::string, my_off_t>;
binlog_position_type last_acked
```

File: plugin/semisync/semisync_master.cc [hunk 10]
```diff
@@ -1196,6 +1207,51 @@ int ReplSemiSyncMaster::setWaitSlaveCount(unsigned int new_value) {
   return function_exit(kWho, result);
 }
 
+void ReplSemiSyncMaster::signal_semi_sync_ack(const char *const log_file,
+                                              const my_off_t log_pos) {
+  const auto acked = std::make_pair(std::string(log_file), log_pos);
+  mysql_mutex_lock(&LOCK_last_acked);
+  if (acked > last_acked) {
```
Comment (percona-ysorokin) [inline plugin/semisync/semisync_master.cc:1216]: In theory, this comparison is not 100% correct.
Binlog names like `master-bin.000001` will be increasing to `master-bin.999999` and then will become `master-bin.1000000`
Lexographical comparison of `master-bin.999999` and `master-bin.1000000` does not produce expected results.
Although I admit that having 1 000 000 is not a typical case.
Just wanted to mention this in the comments.

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Comment by inikep** [comment]
PR was updated according to comments

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request closed without merging**